### PR TITLE
add cookbook for neural net classification

### PR DESCRIPTION
**Pull request by sanuj**
Integration data for this cookbook also keeps on changing.

### CODE REVIEW

File: examples/meta/src/neural_nets/feedforward_net_classification.sg [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+CSVFile f_feats_test("../../data/fm_test_real.dat")
+CSVFile f_labels_train("../../data/label_train_twoclass.dat")
+
+#![create_features]
+RealFeatures features_train(f_feats_train)
+RealFeatures features_test(f_feats_test)
+BinaryLabels labels_train(f_labels_train)
+#![create_features]
+
+#![add_layers]
```
Comment (sanuj) [inline examples/meta/src/neural_nets/feedforward_net_classification.sg]: I tried to do something like this but it was causing problems in translation:
```
NeuralLayers layers()
DynamicObjectArray all_layers =  layers.input(num_feats).linear(50).softmax(2).done()
NeuralNetwork network(all_layers)
```

File: examples/meta/src/neural_nets/feedforward_net_classification.sg [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+layers.append_element(output_layer)
+#![add_layers]
+
+#![create_instance]
+NeuralNetwork network(layers)
+network.quick_connect()
+network.initialize_neural_network()
+#![create_instance]
+
+#![set_parameters]
```
Comment (sanuj) [inline examples/meta/src/neural_nets/feedforward_net_classification.sg]: This is incomplete. If I uncomment the lines below then I get problems in translation. I think cookbook lacks the functionality to assign values to a class member variable.

Comment (karlnapf) [inline examples/meta/src/neural_nets/feedforward_net_classification.sg]: It only allows for SWIG interfaces, so everything that you can do from python, you can do from meta examples

File: examples/meta/src/neural_nets/feedforward_net_classification.sg (hunk 1 above)
Comment (karlnapf) [inline examples/meta/src/neural_nets/feedforward_net_classification.sg]: I assume this works in Python?
In that case, it is the grammar, which doesn't allow subsequent method calls.
Check if it works if you do this in a sequence of steps:
```
DynamicObjectArray all_layers = layers.input(num_feats)
all_layers = all_layers.linear(50)
all_layers = all_layers.softmax(2)
all_layers = all_layers.done()
```
 (or similar)

File: doc/cookbook/source/index.rst [hunk 3]
```diff
@@ -49,3 +49,12 @@ Clustering
...
    examples/clustering/**
+
+Neural Networks
+---------------
+
+.. toctree::
+   :maxdepth: 1
+   :glob:
+
+   examples/neural_nets/**
```
Comment (karlnapf) [inline doc/cookbook/source/index.rst]: good!

File: doc/cookbook/source/references.bib [hunk 4]
```diff
@@ -59,4 +59,11 @@ @book{scholkopf2002learning
   author={B. Sch{\"o}lkopf and A.J. Smola},
   year={2002},
   publisher={MIT press}
+}
+@unpublished{Goodfellow-et-al-2016-Book,
+    title={Deep Learning},
+    author={I. Goodfellow and Y. Bengio and A. Courville},
+    note={Book in preparation for MIT Press},
+    url={http://www.deeplearningbook.org},
```
Comment (karlnapf) [inline doc/cookbook/source/references.bib]: please no URLs in ourbibtex

File: examples/meta/src/neural_nets/feedforward_net_classification.sg [hunk 5]
```diff
@@ -0,0 +1,45 @@
+CSVFile f_feats_train("../../data/fm_train_real.dat")
```
Comment (karlnapf) [inline examples/meta/src/neural_nets/feedforward_net_classification.sg]: can you use the other datasets that have ground truth available?

File: examples/meta/src/neural_nets/feedforward_net_classification.sg (hunk 2 above)
Comment (karlnapf) [inline examples/meta/src/neural_nets/feedforward_net_classification.sg]: yeah you need method calls here. Does this kind of stuff work from swig? If so, we should change the grammar @sorig

File: examples/undocumented/python_modular/neuralnets_simple_modular.py [hunk 6]
```diff
@@ -1,30 +0,0 @@
...
-label_traindat = '../data/label_train_twoclass.dat'
-
-parameter_list = [[traindat,testdat,label_traindat,0.9,1e-3],[traindat,testdat,label_traindat,0.8,1e-2]]
-
-def neuralnets_simple_modular (train_fname, test_fname,
-		label_fname, C, epsilon):
-
-	from modshogun import NeuralLayers, NeuralNetwork, RealFeatures, BinaryLabels
-	from modshogun import Math_init_random, CSVFile
-	Math_init_random(17)
```
Comment (karlnapf) [inline examples/undocumented/python_modular/neuralnets_simple_modular.py]: this example can definitely be translated to meta examples .... nothing that the grammar doesnt support. Just have to flatten the chained method calls

File: examples/meta/src/neural_nets/feedforward_net_classification.sg (hunk 1 above)
Comment (sanuj) [inline examples/meta/src/neural_nets/feedforward_net_classification.sg]: @karlnapf Even after the changes by @lisitsyn, it gives segfault.
```
auto num_feats = features_train->get_num_features();
auto layers = some<CNeuralLayers>();
layers = layers->input(num_feats);
layers = layers->linear(50);
layers = layers->softmax(2);
auto all_layers = layers->done();
```
Above is the `cpp` translation. Segmentation fault occurs in the 4th line.

Comment (lisitsyn) [inline examples/meta/src/neural_nets/feedforward_net_classification.sg]: Lacks `%newobject`

Comment (karlnapf) [inline examples/meta/src/neural_nets/feedforward_net_classification.sg]: @sanuj check the swig interface file (dont know how it is called, but you need to tell swig that a method returns a new object in the occresponding `.i` file)

**Comment by sanuj** [comment]
@karlnapf why [this](https://travis-ci.org/shogun-toolbox/shogun/jobs/140307872#L1678) error? is there a bug in the `.cs` generation?

File: examples/meta/src/neural_nets/feedforward_net_classification.sg [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+#![set_parameters]
+
+#![train_and_apply]
+network.set_labels(labels_train)
+network.train(features_train)
+BinaryLabels labels_predict = network.apply_binary(features_test)
+#![train_and_apply]
+
+#![get_params]
+RealVector params = network.get_parameters()
```
Comment (karlnapf) [inline examples/meta/src/neural_nets/feedforward_net_classification.sg]: maybe call this different than params and see what csharp says

**Comment by karlnapf** [comment]
@sanuj can you paste the generated csharp listing as a gist? There must be a problem in there, not sure why without seeing it. Might be some reserved keyword....

**Comment by sanuj** [comment]
@karnlapf where can i find the csharp listing? Or shall i just change the name of 'params'?

**Comment by karlnapf** [comment]
You run `make meta_examples` and check your build folder locally

**Comment by karlnapf** [comment]
I dont know whether that is the problem, you have to check and play with it

**Comment by sanuj** [comment]
@karlnapf review and merge.

File: examples/meta/src/neural_nets/feedforward_net_classification.sg [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+network.train(features_train)
+BinaryLabels labels_predict = network.apply_binary(features_test)
+#![train_and_apply]
+
+#![get_params]
+RealVector parameters = network.get_parameters()
+#![get_params]
+
+#![evaluate_accuracy]
+AccuracyMeasure eval()
```
Comment (karlnapf) [inline examples/meta/src/neural_nets/feedforward_net_classification.sg]: minor, pls dont call this "eval" 
Can send a mini pr to fix

**Comment by karlnapf** [comment]
Nice one!

**Pull request merged**